File: coek/models/misc/pmedian.cpp
#include <map>
#include <vector>
#include <coek/coek.hpp>


void pmedian_scalar(coek::Model& model, const std::vector<int>& data)
{
if (data.size() != 2)
    throw std::runtime_error("pmedian_scalar - expecting two arguments (N,P)");
int N = data[0];   // Locations
int P = data[1];   // Facilities

int M = N;          // Customers

std::vector<std::vector<double>> d(N, std::vector<double>(M));
for (int n=0; n<N; n++)
    for (int m=0; m<M; m++)
        d[n][m] = 1.0+1.0/(n+m+1);

std::vector<std::vector<coek::Variable>> x(N, std::vector<coek::Variable>(M));
for (int n=0; n<N; n++)
    for (int m=0; m<M; m++)
        model.add( x[n][m].bounds(0,1).value(0) );

std::vector<coek::Variable> y(N);
for (int n=0; n<N; n++)
    model.add( y[n].bounds(0,1).value(0) );

// obj
coek::Expression obj;
for (int n=0; n<N; n++)
    for (int m=0; m<M; m++)
        obj += d[n][m]*x[n][m];
model.add_objective( obj );

// single_x
for (int m=0; m<M; m++) {
    coek::Expression c;
    for (int n=0; n<N; n++)
        c += x[n][m];
    model.add( c == 1 );
    }

// bound_y
for (int n=0; n<N; n++)
    for (int m=0; m<M; m++)
        model.add( x[n][m] - y[n] <= 0 );

// num_facilities
coek::Expression num_facilities;
for (int n=0; n<N; n++)
    num_facilities += y[n];
model.add( num_facilities == P );
}

File: tmp/aml_comparisons/pyomo/models/pmedian.py
import pyomo.environ as pe


def pmedian(N, P):

    model = pe.ConcreteModel()

    model.N = N

    model.P = P

    model.M = N

    model.Locations = pe.RangeSet(1,model.N)

    model.Customers = pe.RangeSet(1,model.M)

    model.d = pe.Param(model.Locations, model.Customers, initialize=lambda model, n, m : 1.0+1.0/(n+m+1), within=pe.Reals)

    model.x = pe.Var(model.Locations, model.Customers, bounds=(0.0,1.0), initialize=0.0)

    model.y = pe.Var(model.Locations, bounds=(0.0, 1.0), initialize=0.0)

    def rule(model):
        return sum( model.d[n,m]*model.x[n,m] for n in model.Locations for m in model.Customers )
    model.obj = pe.Objective(rule=rule)

    def rule(model, m):
        return sum( model.x[n,m] for n in model.Locations ) == 1.0
    model.single_x = pe.Constraint(model.Customers, rule=rule)

    def rule(model, n,m):
        return model.x[n,m] <= model.y[n]
    model.bound_y = pe.Constraint(model.Locations, model.Customers, rule=rule)

    def rule(model):
        return sum( model.y[n] for n in model.Locations ) - model.P == 0.0
    model.num_facilities = pe.Constraint(rule=rule)

    return model

